serialize/src/lib.rs

```diff
@@ -94,6 +96,44 @@ pub trait CanonicalSerialize {
     }
 }
 
+// This private struct works around Serialize taking the pre-existing
+// std::io::Write instance of most digest::Digest implementations by value
+struct HashMarshaller<'a, H: Digest>(&'a mut H);
+
+impl<'a, H: Digest> ark_std::io::Write for HashMarshaller<'a, H> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> ark_std::io::Result<usize> {
+        Digest::update(self.0, buf);
+        Ok(buf.len())
+    }
+
+    #[inline]
+    fn flush(&mut self) -> ark_std::io::Result<()> {
+        Ok(())
+    }
+}
+
+/// The CanonicalSerialize induces a natural way to hash the
+/// corresponding value, of which this is the convenience trait.
+pub trait CanonicalSerializeExt: CanonicalSerialize {
```

Comment:
Maybe we call this something like `CanonicalSerializeThenHash`?

Reply:
- The serialization is never materialized anywhere, whereas `CanonicalSerializeThenHash` implies that it is, in the way the peudo-code in #86 did.
- The best practice for an extension trait is really to finish in `Ext`. `CanonicalSerializeHashExt`?

Reply:
`CanonicalSerializeHashExt` sounds good.